src/compiler/crystal/tools/doc/json_serializer.cr

```diff
@@ -0,0 +1,175 @@
+require "json"
+require "json/*"
+
+module Crystal::Doc::JSONSerializer
+  macro mapping(src, properties, strict = false)
+    JSON.mapping({{properties}}, {{strict}})
+
+    def initialize(src : {{src.id}})
+      {% for key, value in properties %}
+        {% if value[:wrap] %}
+          {% if value[:type].is_a?(Generic) %}
```

Comment:
This assumes that all `Generic` types are `Array(T)`. Wouldn't it be better to check this assumption? It also seems you use the `wrap` property to get the generic type. You should be able to use `type_vars[0]` to get this from the `generic`.